### PR TITLE
MGMT-4170 - Create assisted service python client as a library

### PR DESCRIPTION
**Pull request by eliorerz**
Create assisted service python client as a library , make generate-python-client now creates python package

### CODE REVIEW

**Comment by eliorerz** [comment]
@YuviGold

File: tools/client_package_initializer.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+        for line in lines:
+            if "setup(" in line:
+                data += "SETUP = dict(\n"
+            else:
+                data += line
+
+        self._setup_data.update(self._execute(data)["SETUP"])
+        self._setup_data['author'] = "RedHat"
+        self._setup_data['author_email'] = "UNKNOWN"
```
Comment (YuviGold) [inline tools/client_package_initializer.py:92]: Please set those fields on the `__init__` function. No reason it wouldn't be updated using the `update` function.

File: tools/client_package_initializer.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        data = ""
+
+        with open(self._setup_path) as f:
+            lines = f.readlines()
+
+        for line in lines:
+            if "setup(" in line:
+                data += "SETUP = dict(\n"
+            else:
+                data += line
```
Comment (YuviGold) [inline tools/client_package_initializer.py:88]: I don't understand what's going on over here.. seems like an hack :(
Isn't that possible that that setup.py includes other not-dict related lines?
In addition, when do you stop loading lines? what happens if `setup(` doesn't exist? Is that naive loading just works?

File: tools/client_package_initializer.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        self._project_path = project_path
+        self._setup_path = os.path.join(project_path, self.DEFAULT_SETUP_PATH)
+        self._url = url
+        self._version = None
+        self._setup_data = {"setup_requires": ['vcversioner'],
+                            "vcversioner": {'vcs_args': ['git', 'describe', '--tags', '--long']}}
+        self._tests_require = ""
+
+        with open("MANIFEST.in", "w") as f:
+            f.write("include README.md")
```
Comment (YuviGold) [inline tools/client_package_initializer.py:73]: f.write(f"include {self.DEFAULT_README_PATH}")

File: hack/generate.sh [hunk 4]
```diff
@@ -26,21 +26,17 @@ function generate_go_client() {
...
+    rm -rf "${__root}"/assisted-service-client*
+
+    docker run --rm -u "$(id -u)" --entrypoint /bin/sh \
+        -v "${__root}":/local:Z \
+        -v "${__root}"/swagger.yaml:/swagger.yaml:ro,Z \
+        -v "${__root}"/tools/generate_python_client.sh:/script.sh:ro,Z \
         -e SWAGGER_FILE=/swagger.yaml -e OUTPUT=/local/assisted-service-client/ \
         swaggerapi/swagger-codegen-cli:2.4.15 /script.sh
-    cd ${dest}/assisted-service-client/ && python3 setup.py sdist --dist-dir ${dest}
+     cp "${__root}"/tools/client_package_initializer.py "${__root}"/assisted-service-client/client_package_initializer.py
```
[the remark this answers is not in the document]
Comment (eliorerz) [reply hack/generate.sh:37]: It creates a lot of unnecessary files during the build process. It more organized that all are in the place.

[the remark this answers is not in the document]
Comment (YuviGold) [reply hack/generate.sh:37]: Can you create them relative to the destination location?

File: tools/client_package_initializer.py (hunk 2 above)
Comment (eliorerz) [reply tools/client_package_initializer.py:88]: @YuviGold  The signature of setup is `def setup(**attrs)` so by definition it's acting like python dict.
My assumption was that if the  `setup( ` keyword is missing there was a problem generating client code and we didn't need to proceed the building process .
There is one library that does the same (parsing setup.py) but it also uses hack and assumptions on setup.py file.
We can discuss this issue offline if you will like.

**Review by YuviGold — APPROVED** [review]
LGTM

**Comment by YuviGold** [comment]
Passed in Jenkins :) http://assisted-jenkins.usersys.redhat.com/job/assisted-service/job/PR-1168/36/
/unhold

**Review by YuviGold — APPROVED** [review]
LGTM

**Pull request merged**